pypsa/components/abstract.py

```diff
@@ -0,0 +1,252 @@
+from pypsa.definitions.structures import Dict
+from pypsa.utils import equals
+
+logger = logging.getLogger(__name__)
+
+if TYPE_CHECKING:
+    from pypsa import Network
+
+# TODO attachment todos
+# - crs
+# - snapshots, investment_periods
+
+
+@dataclass
+class ComponentsData:
+    ct: ComponentTypeInfo
+    n: Network | None
+    static: pd.DataFrame
+    dynamic: dict
+
+
+class Components(ComponentsData, ABC):
+    """
+    Abstract base class for Container of energy system related assets, such as
+    generators or transmission lines. Use the specific subclasses for concrete or
+    a generic component type.
+
+    .. warning::
+        This class is under ongoing development and will be subject to changes.
+        It is not recommended to use this class outside of PyPSA.
+    """
+
+    def __init__(
+        self,
+        ct: ComponentTypeInfo,
+        n: Network | None = None,
+        names: str | int | Sequence[int | str] | None = None,
+        suffix: str = "",
+    ) -> None:
+        if names is not None:
+            msg = "Adding components during initialisation is not yet supported."
+            raise NotImplementedError(msg)
+        if n is not None:
+            msg = (
+                "Attaching components to Network during initialisation is not yet "
+                "supported."
+            )
+            raise NotImplementedError(msg)
+        static, dynamic = self._get_data_containers(ct)
+        super().__init__(ct, n=None, static=static, dynamic=dynamic)
+
+    def __repr__(self) -> str:
+        num_components = len(self.static)
+        if not num_components:
+            return f"Empty PyPSA {self.ct.name} Components\n"
+        text = f"PyPSA '{self.ct.name}' Components"
+        text += "\n" + "-" * len(text) + "\n"
+
+        # Add attachment status
+        if self.attached:
+            network_name = f"'{self.n_save.name}'" if self.n_save.name else ""
+            text += f"Attached to PyPSA Network {network_name}\n"
+
+        text += f"Components: {len(self.static)}"
+
+        return text
+
+    def __getitem__(self, key: str) -> Any:
+        return getattr(self, key)
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        if key in self.__dict__:
+            setattr(self, key, value)
+        else:
+            raise KeyError(f"'{key}' not found in Component")
+
+    def __eq__(self, other: Any) -> bool:
+        return (
+            equals(self.ct, other.ct)
+            and equals(self.static, other.static)
+            and equals(self.dynamic, other.dynamic)
+        )
+
+    @staticmethod
+    def _get_data_containers(ct: ComponentTypeInfo) -> tuple[pd.DataFrame, Dict]:
+        static_dtypes = ct.defaults.loc[ct.defaults.static, "dtype"].drop(["name"])
+        if ct.name == "Shape":
+            # TODO: Needs general default crs
+            crs = CRS.from_epsg(
+                "4326"
+            )  # if n is None else n.crs #TODO attach mechanism
```

Comment:
4326 is good default

Reply:
Yes, the TODO just means it would be better to define the default somewhere and import it, instead of defining it multiple times (in components and network). Done now via `constants.py`.